hyperspy/axes.py

```diff
@@ -974,17 +953,20 @@ def __init__(self,
         for parameter in parameters.keys():
             self.add_trait(parameter, t.CFloat(parameters[parameter]))
         self.parameters_list = list(parameters.keys())
-        self.update_axis()
-        self.on_trait_change(self.update_axis, self.parameters_list)
 
-    def update_axis(self):
+    @property
+    def axis(self):
```

Comment:
It may be worth caching it in case the function calculation is slow.

Reply:
Okay so I think the way to do this if we want to go down this path would be to keep the `_axis` attribute for all axes.

Then we could define a function like this:

```python
@observe('scale', postinit=True)
@observe('offset', postinit=True)
@observe('size', postinit=True)
def update_axis(event):
    self._axis  = np.arange(self.size)*self.scale+self.offset
```

Then for the FunctionalDataAxis we could create a function

```python
@observe(trait("x._axis", notify=False).list_items(), postinit=True)
@observe(self.parameters_list, postinit=True)
def update_axis(event):
    kwargs = {}
    for kwarg in self.parameters_list:
        kwargs[kwarg] = getattr(self, kwarg)
    new_axis = self._function(x=self.x.axis, **kwargs)
    # Set not valid values to np.nan
    new_axis[np.logical_not(np.isfinite(new_axis))] = np.nan
    self._axis = new_axis
```

Reply:
Part of the reason I don't really like this is the fact that even though the `._axis` property is protected you could still directly modify the values for a UniformDataAxis and FunctionalDataAxis

Reply:
After looking through the documentation for Traits I think it should be fairly easy to cache the property.  I hadn't looked into the additional capabilities for caching.


```python
from traits.api import cached_property, Property

axis =  Property(observe=['scale", "size", "offset"])


@cached_property
def _get_axis(self):
    return np.arange(self.size)*self.scale+self.offset

```

https://docs.enthought.com/traits/traits_user_manual/advanced.html#caching-a-property-value

Reply:
Or maybe it isn't quite this easy... Nothing with the traits package has been quite as intuitive as I first thought....

Reply:
@ericpre So there are a few problems with this.  Mostly that inplace editing of traits is not supported.  We have been operating by adding traits to each object individually rather than adding traits to the Class as a whole which is not well supported for . The problem is that when we change the `__class__` variable and call `__init__` we skip over the `__new__` function where the listeners are set for some class. 

Unfortunately I think that means that in order to cache axes we would have copy axes when converting from one type to the next. 

That means that `s.axes_manager[0].convert_to_nonuniform_axis()` --> `s.axes_manager[0] = s.axes_manager[0].convert_to_nonuniform_axis()`

It isn't a huge change but would allow us to define the Axes classes like this.

```python
from traits.api import (cached_property, HasTraits, Str,Float, Int,observe, Property,Array, Undefined)

class BaseDataAxis(HasTraits):
    name=Str(Undefined)
    units=Str(Undefined)  

class DataAxis(BaseDataAxis):
    axis = Array()

class FunctionalDataAxis(DataAxis):
    x = Instance(BaseDatatAxis)
    axis = Property(observe=["x.axis"] # overwrites axis in DataAxis

    @cached_property
    def _get_axis(self):
        print("calculating axis")
        return np.arange(self.size)*self.scale+self.offset

class DataAxis(DataAxis):
    scale = Float(1)
    offset = Float(0)
    size = Int(1)
    axis = Property(observe=["scale", "offset", "size"] # overwrites axis in DataAxis

    @cached_property
    def _get_axis(self):
        print("calculating axis")
        return np.arange(self.size)*self.scale+self.offset
```